torch/_functorch/aot_autograd.py

```diff
@@ -970,6 +998,35 @@ def functionalized_f_helper(primals_and_maybe_tangents):
         finally:
             torch._disable_functionalization()
 
+        if keep_input_mutations:
+            # Note: This is a bit annoying. There's a layering issue here, where:
+            # (1) functionalization needs to operate on **synthetic base** inputs, before unpacking them into the "real" inputs.
+            # (2) For keep_input_mutations, we support tracing a call to copy_() directly on mutated inputs.
+            #     However, we **only** want to support this for inputs that have data-only (and no metadata) mutations,
+            #     because inductor (and backends in generally) would prefer not to see these (e.g. as_strided_(), resize_()).
+            #     This makes it pretty difficult for this logic to operate on synthetic bases.
+            #     For example, suppose "x" and "y" are aliased inputs, and we witness:
+            #       x.add_(1)
+            #       y.t_()
+            #     We don't actually want the compiler to witness a copy_() on the synthetic base here: it should
+            #     just see the copy_() directly on x!
+            #     For now, to do this, I'm manually unpacking the synthetic bases again in this function.
+            primals_unpacked = unpack_synthetic_bases(primals, synthetic_base_info)
+            f_primals_unpacked = unpack_synthetic_bases(f_primals, synthetic_base_info)
+            assert len(meta.input_info) == len(f_primals_unpacked)
+            for i, (inpt_old, inpt_f) in enumerate(zip(primals_unpacked, f_primals_unpacked)):
+                if not isinstance(inpt_f, torch.Tensor):
+                    continue
+                torch._sync(inpt_f)
+                inpt_new = torch._from_functional_tensor(inpt_f)
+                if meta.input_info[i].mutates_data and not meta.input_info[i].mutates_metadata:
+                    # We found an input that had a (data-only) mutation.
+                    # Since keep_input_mutations is set, we need to faithfully apply a copy_()
+                    # so the compiler will see the input mutation in the graph.
+                    assert inpt_new is not inpt_old
+                    assert has_same_metadata(inpt_new, inpt_old)
+                    inpt_old.copy_(inpt_new)
```

Comment:
What I was hoping to see here, code structure wise, was a function that "applies copy_" and it being called from both the runtime wrapper, or here, guarded by the `keep_input_mutation` flag which controlled where you ran it. This would be make it obvious that the same logic was being used in the wrapper as in codegen. On the plus side I suppose this code is not that long. But it is not at all clear from the code itself (without your PR comment in https://github.com/pytorch/pytorch/pull/92857/files/6bdf2563239b88ce9dbbe52afbde3e37a2e24412..c764f7fa0478e24176bc422888a44ae391d1b1fa#r1095151376 which will be lost to the sands of time when this PR lands) that this is what's going on

Reply:
> What I was hoping to see here, code structure wise, was a function that "applies copy_" and it being called from both the runtime wrapper, or here

Definitely agreed that a shared function like this, even if it doesn't deduplicate much code, would make things clearer (you can read the code and understand that the copy_() calls happen either at runtime or trace time, gated on a flag).

Definitely open to feedback - but the main reason I didn't end up trying to refactor in that way was perf. Today, at trace time we pre-compute a set of "input indices" that we need to loop over in the runtime epilogue, to handle any data / metadata mutations (`runtime_metadata.fw_metadata.mutated_inp_indices`). I opted to pre-compute this list to **not**  include the data-only mutations when `keep_inference_input_mutations` is set ([code comment](https://github.com/pytorch/pytorch/pull/92857/files#diff-df954bbf954d2dcb81f687876053267ffa4ddb36ed86b7d2bd76319ff2b94416R446)). Concretely, if we we compile an optimizer step with 100 data-only input mutations, then this will avoid looping over 100 indices that we no-op on.

Either way, for sure this set of conventions and assumptions in the code is complicated. My thought is that is that we want both better code clarity and we don't want to sacrifice runtime overhead, so I can add some clearer comments (I have some already, that can probably be improved).

Reply:
I buy the precompute argument. Leave it as is.

Reply:
If we really really wanted it to happen, you'd probably have to arrange for the precompute to happen at compile time too, so you can call the precompute copy apply.